src/python/pants/engine/internals/dep_rules.py

```diff
@@ -0,0 +1,202 @@
+from __future__ import annotations
+
+import logging
+from abc import ABC, abstractmethod
+from dataclasses import dataclass, field
+from enum import Enum
+from typing import Iterable, Mapping, Sequence, Tuple, Union
+
+from typing_extensions import Literal
+
+from pants.engine.internals.target_adaptor import TargetAdaptor
+from pants.engine.rules import Get, collect_rules, rule
+from pants.engine.unions import UnionMembership, union
+
+logger = logging.getLogger(__name__)
+
+
+SetDependencyRulesValueT = Tuple[str, ...]
+SetDependencyRulesKeyT = Union[str, Tuple[str, ...]]
+SetDependencyRulesT = Mapping[SetDependencyRulesKeyT, SetDependencyRulesValueT]
+
+
+class DependencyRulesError(Exception):
+    pass
+
+
+class DependencyRuleActionDeniedError(DependencyRulesError):
+    pass
+
+
+class DependencyRuleAction(Enum):
+    ALLOW = "allow"
+    DENY = "deny"
+    WARN = "warn"
+
+    def execute(self, *, description_of_origin: str) -> None:
+        if self is DependencyRuleAction.ALLOW:
+            return
+        msg = f"Dependency rule violation for {description_of_origin}"
+        if self is DependencyRuleAction.DENY:
+            raise DependencyRuleActionDeniedError(msg)
+        if self is DependencyRuleAction.WARN:
+            logger.warning(msg)
+
+
+SetDefaultDependencyRulesT = Union[Literal["allow"], Literal["deny"], Literal["warn"]]
+
+
+class DependencyRule:
+    pass
+
+
+DependencyRules = Tuple[DependencyRule, ...]
+
+
+class BuildFileDependencyRules(ABC):
+    default: DependencyRuleAction
+    all: DependencyRules
+    targets: Mapping[str, DependencyRules]
+
+    @classmethod
+    @abstractmethod
+    def create(
+        cls,
+        default: DependencyRuleAction = DependencyRuleAction.ALLOW,
+        all: Iterable[str | DependencyRule] = (),
+        targets: Mapping[str, Iterable[str | DependencyRule]] = {},
+    ) -> BuildFileDependencyRules:
+        ...
+
+    @staticmethod
+    @abstractmethod
+    def check_dependency_rules(
+        *,
+        source_adaptor: TargetAdaptor,
+        source_path: str,
+        dependencies_rules: BuildFileDependencyRules | None,
+        target_adaptor: TargetAdaptor,
+        target_path: str,
+        dependents_rules: BuildFileDependencyRules | None,
+    ) -> DependencyRuleAction:
+        """The source of the dependency has the dependencies field, the target of the dependency is
+        the one listed as a value in the dependencies field.
+
+        The `__dependencies_rules__` are the rules applicable for the source path.
+        The `__dependents_rules__` are the rules applicable for the target path.
+
+        Return dependency rule action ALLOW, DENY or WARN. WARN is effectively the same as ALLOW,
+        but with a logged warning.
+        """
+
+
+@dataclass
+class BuildFileDependencyRulesParserState:
+    parent: BuildFileDependencyRules | None
+    default: DependencyRuleAction = DependencyRuleAction.ALLOW
```

Comment:
Random thought:

To simplify the specification and the parsing, maybe instead of `all` and `default` being special keywords, we allow wildcards in the type_alias?

So `python_*` matches `python_sources` and `python_tests`, and `*` matches everything (and so replaces `all`)? 

And, now that I think of it, do we need `default` if we have `all`/`*`, since we also have wildcards for the paths?

Reply:
Done! Thanks :)